### PR TITLE
roachtest: add asyncpg test suite

### LINKED ISSUE #70209: roachtest: run asyncpg test suite in nightlies
I ran the `asyncpg` test suite locally with the following commands:

```
git clone --recurse-submodules git@github.com:MagicStack/asyncpg.git
cd asyncpg
pip install -e .
# modified asyncpg/connect_utils.py to point at CRDB
python setup.py test
```

It passed except for two tests that check the Postgres port
```
======================================================================
FAIL: test_connect_params (test_connect.TestConnectParams)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/rafiss/scratch/asyncpg/tests/test_connect.py", line 792, in test_connect_params
    self.run_testcase(testcase)
  File "/Users/rafiss/scratch/asyncpg/tests/test_connect.py", line 748, in run_testcase
    self.assertEqual(expected, result, 'Testcase: {}'.format(testcase))
AssertionError: Tuples differ: ([('host1', 5432), ('host2', 5432)], {'database': 'db', 'user': 'user'}) != ([('host1', 26257), ('host2', 26257)], {'user': 'user', 'database': 'db'})

First differing element 0:
[('host1', 5432), ('host2', 5432)]
[('host1', 26257), ('host2', 26257)]

- ([('host1', 5432), ('host2', 5432)], {'database': 'db', 'user': 'user'})
?              ^^^              ^^^

+ ([('host1', 26257), ('host2', 26257)], {'database': 'db', 'user': 'user'})
?             +++ ^             +++ ^
 : Testcase: {'name': 'dsn_only_multi_host', 'dsn': 'postgresql://user@host1,host2/db', 'result': ([('host1', 5432), ('host2', 5432)], {'database': 'db', 'user': 'user'})}

======================================================================
FAIL: test_connect_pgpass_regular (test_connect.TestConnectParams)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/rafiss/scratch/asyncpg/tests/test_connect.py", line 845, in test_connect_pgpass_regular
    self.run_testcase({
  File "/Users/rafiss/scratch/asyncpg/tests/test_connect.py", line 748, in run_testcase
    self.assertEqual(expected, result, 'Testcase: {}'.format(testcase))
AssertionError: Tuples differ: ([('abc', 5432)], {'password': 'password from pgpas[46 chars]db'}) != ([('abc', 26257)], {'user': 'user', 'password': 'pa[47 chars]db'})

First differing element 0:
[('abc', 5432)]
[('abc', 26257)]

- ([('abc', 5432)],
?            ^^^

+ ([('abc', 26257)],
?           +++ ^

   {'database': 'db',
    'password': 'password from pgpass for user@abc',
    'user': 'user'}) : Testcase: {'dsn': 'postgres://user@abc/db?passfile=/var/folders/1_/_kcb28_94w16ml515w877ynm0000gn/T/tmpjh5aawll', 'result': ([('abc', 5432)], {'password': 'password from pgpass for user@abc', 'user': 'user', 'database': 'db'})}

----------------------------------------------------------------------
Ran 284 tests in 122.763s

FAILED (failures=2, skipped=3)
Test failed: <unittest.runner.TextTestResult run=284 errors=0 failures=2>
error: Test failed: <unittest.runner.TextTestResult run=284 errors=0 failures=2>
```

Let's turn this into a nightly test that automates all the above steps.

### PR DESCRIPTION
**Pull request by ZhouXing19**
Fix #70209.

Add a new roachtest, asyncpg, that changes the connection port
to point at cockroachdb, and runs the original test suite from
the asyncpg package (`python3 setup.py test`).

Release note: None.

### CODE REVIEW

File: pkg/cmd/roachtest/cluster.go [hunk 1]
```diff
@@ -1737,6 +1737,26 @@ fi
 		branch))
 }
 
+// GitCloneWithRecurseSubmodules clones a git repo from src into dest and checks out origin's
+// version of the given branch, but with a --recurse-submodules flag.
+// The src, dest, and branch arguments must not contain shell special characters.
+func (c *clusterImpl) GitCloneWithRecurseSubmodules(
```
[the remark this answers is not in the document]
Comment (ZhouXing19) [reply pkg/cmd/roachtest/cluster.go]: @tbg TFTR! Moved it to `tests/canary.go`.

File: pkg/cmd/roachtest/tests/asyncpg.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+	"context"
+	"fmt"
+
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/registry"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/spec"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test"
+)
+
+func registerAsyncpg(r registry.Registry) {
```
Comment (otan) [inline pkg/cmd/roachtest/tests/asyncpg.go:23]: nit: to make this Go style compliant, we should call everything `AsyncPG` (acronyms are always the same case, capital letters if it is not the first word, e.g. pgCompat, compatPG)

File: pkg/cmd/roachtest/tests/asyncpg.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+			t.Fatal(err)
+		}
+
+		if err := repeatRunE(
+			ctx,
+			t,
+			c,
+			node,
+			"install postgresql",
+			`sudo apt-get -y install postgresql`,
```
Comment (otan) [inline pkg/cmd/roachtest/tests/asyncpg.go:77]: curious, why do we need this? can you add a comment?

File: pkg/cmd/roachtest/tests/canary.go [hunk 4]
```diff
@@ -291,3 +292,27 @@ func repeatGetLatestTag(
...
+) error {
+	return c.RunL(ctx, l, node, "bash", "-e", "-c", fmt.Sprintf(`'
+if ! test -d %s; then
+  git clone --recurse-submodules -b %s --depth 1 %s %s
+else
+  cd %s
+  git fetch origin
+  git checkout origin/%s
+fi
+'`, dest,
```
Comment (otan) [inline pkg/cmd/roachtest/tests/canary.go:314]: let's use `%[1]s, %[2]s, ...` here to save duplicate arguments.

```
`if ! test -d %[1]s; then
  git clone --recurse-submodules -b %[2]s --depth 1 %[3]s %[1]s
else
  cd %[1]s
  git fetch origin
  git checkout origin/%[2]s
fi`,
dest,
branch,
src,
```

File: pkg/cmd/roachtest/tests/asyncpg.go (hunk 2 above)
Comment (otan) [reply pkg/cmd/roachtest/tests/asyncpg.go:23]: hmm, actually maybe ignore this -- it's stylised asyncpg /shrug

**Review by ZhouXing19 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71029)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan, @tbg, and @ZhouXing19)

___
*[pkg/cmd/roachtest/tests/asyncpg.go, line 77 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/71029#-MkzFG1HtDNRmlUjfPR6-r2-77:-MlL34P60eeYaok0jAyg:bp36pvl) ([raw file](https://github.com/cockroachdb/cockroach/blob/105cd1261f19a360db9e68cbafa51e093a333ec6/pkg/cmd/roachtest/tests/asyncpg.go#L77)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

curious, why do we need this? can you add a comment?
</blockquote></details>

It's because asyncpg requires a `pg_ctl` for the setup. See [asyncpg's code here](https://github.com/MagicStack/asyncpg/blob/383c711eb68bc6a042c121e1fddfde0cdefb8068/asyncpg/cluster.py#L407-L420).

Without this PostgreSQL installation, `roachtest run asyncpg` gives:
```
Traceback (most recent call last):
  File "/mnt/data1/asyncpg/asyncpg/_testbase/__init__.py", line 319, in setUpClass
    cls.setup_cluster()
  File "/mnt/data1/asyncpg/tests/test_connect.py", line 1222, in setup_cluster
    cls.cluster = cls.new_cluster(pg_cluster.TempCluster)
  File "/mnt/data1/asyncpg/asyncpg/_testbase/__init__.py", line 298, in new_cluster
    cluster = _init_cluster(ClusterCls, cluster_kwargs,
  File "/mnt/data1/asyncpg/asyncpg/_testbase/__init__.py", line 211, in _init_cluster
    cluster.init(**(initdb_options or {}))
  File "/mnt/data1/asyncpg/asyncpg/cluster.py", line 109, in init
    if self.get_status() != 'not-initialized':
  File "/mnt/data1/asyncpg/asyncpg/cluster.py", line 77, in get_status
    self._init_env()
  File "/mnt/data1/asyncpg/asyncpg/cluster.py", line 417, in _init_env
    self._pg_ctl = self._find_pg_binary('pg_ctl')
  File "/mnt/data1/asyncpg/asyncpg/cluster.py", line 554, in _find_pg_binary
    raise ClusterError(
asyncpg.cluster.ClusterError: could not find pg_ctl executable: '/usr/lib/postgresql/12/bin/pg_ctl' does not exist or is not a file
```
___
*[pkg/cmd/roachtest/tests/canary.go, line 314 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/71029#-MlAvUmWTw57ZaMUb0ri-r2-314:-MlL2ZTW4tZte1tBDr3p:b70mqqs) ([raw file](https://github.com/cockroachdb/cockroach/blob/105cd1261f19a360db9e68cbafa51e093a333ec6/pkg/cmd/roachtest/tests/canary.go#L314)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

let's use `%[1]s, %[2]s, ...` here to save duplicate arguments.

```
`if ! test -d %[1]s; then
  git clone --recurse-submodules -b %[2]s --depth 1 %[3]s %[1]s
else
  cd %[1]s
  git fetch origin
  git checkout origin/%[2]s
fi`,
dest,
branch,
src,
```
</blockquote></details>

Thanks! TIL.

**Review by otan — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71029)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan, @tbg, and @ZhouXing19)

___
*[pkg/cmd/roachtest/tests/asyncpg.go, line 77 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/71029#-MkzFG1HtDNRmlUjfPR6-r2-77:-MlgdfuJ2zDh-SD9jxR5:b-dik1fk) ([raw file](https://github.com/cockroachdb/cockroach/blob/105cd1261f19a360db9e68cbafa51e093a333ec6/pkg/cmd/roachtest/tests/asyncpg.go#L77)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

It's because asyncpg requires a `pg_ctl` for the setup. See [asyncpg's code here](https://github.com/MagicStack/asyncpg/blob/383c711eb68bc6a042c121e1fddfde0cdefb8068/asyncpg/cluster.py#L407-L420).

Without this PostgreSQL installation, `roachtest run asyncpg` gives:
```
Traceback (most recent call last):
  File "/mnt/data1/asyncpg/asyncpg/_testbase/__init__.py", line 319, in setUpClass
    cls.setup_cluster()
  File "/mnt/data1/asyncpg/tests/test_connect.py", line 1222, in setup_cluster
    cls.cluster = cls.new_cluster(pg_cluster.TempCluster)
  File "/mnt/data1/asyncpg/asyncpg/_testbase/__init__.py", line 298, in new_cluster
    cluster = _init_cluster(ClusterCls, cluster_kwargs,
  File "/mnt/data1/asyncpg/asyncpg/_testbase/__init__.py", line 211, in _init_cluster
    cluster.init(**(initdb_options or {}))
  File "/mnt/data1/asyncpg/asyncpg/cluster.py", line 109, in init
    if self.get_status() != 'not-initialized':
  File "/mnt/data1/asyncpg/asyncpg/cluster.py", line 77, in get_status
    self._init_env()
  File "/mnt/data1/asyncpg/asyncpg/cluster.py", line 417, in _init_env
    self._pg_ctl = self._find_pg_binary('pg_ctl')
  File "/mnt/data1/asyncpg/asyncpg/cluster.py", line 554, in _find_pg_binary
    raise ClusterError(
asyncpg.cluster.ClusterError: could not find pg_ctl executable: '/usr/lib/postgresql/12/bin/pg_ctl' does not exist or is not a file
```
</blockquote></details>

hmm, not a big fan of it starting postgres to execute its tests. i wish we could comment it out. oh well.
it needs pg_ctl AND postgres to setup from what I can read. Can you link the code snippet in your comment?

**Review by rafiss — APPROVED** [review]
lgtm! but needs one more `make bazel-generate`

**Pull request merged**